File: 8-string-to-integer-atoi/string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        if (s.length() == 0) {
            return 0;
        }
        int i = 0;
        int sign = 1;
        int result = 0;
        while (i < s.length() && s.charAt(i) == ' ') //Skip the Spaces
        {
            i++;
        }
        //The Sign of the Number is calculated
        if (i < s.length() && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            if(s.charAt(i)=='-')
                sign=-1;
            else
                sign=1;
            i++;
        }
        //Now We find the Number
        while (i < s.length() && Character.isDigit(s.charAt(i))) {
            int digit = s.charAt(i) - '0';
            if (result > (Integer.MAX_VALUE - digit) / 10) {
                if(sign==1)
                    return Integer.MAX_VALUE;
                else
                    return Integer.MIN_VALUE;
            }
            result = result * 10 + digit;
            i++;
        }
        return result * sign;
    }
}

File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int convert(int idx, string s, int number, int sign) {
    
        if ( idx == s.length() || !(s[idx] >= '0' && s[idx] <= '9') ) {
            return number * sign;
        }
        int digit = s[idx] - '0';
        
        if ( number > (INT_MAX - digit)/10 ) {
            if (sign == 1) {
                return INT_MAX;
            } else {
                return INT_MIN;
            }
        }
        number = (number * 10) + digit;
        // cout << number << endl;
        number = convert(idx + 1, s, number, sign);

        return number;        
    }
    int myAtoi(string s) {
        int idx = 0;
        while (s[idx] == ' ') {
            idx++;
        }

        int sign = 0;
        int ans = 0;
        if (s[idx] == '-') {
            sign = -1;
            idx++;
        } else {
            sign = 1;
            if (s[idx] == '+') idx++;
        }
        // cout << sign << " " << idx << " idx" << endl;

        int number = 0;
        number = convert(idx, s, 0, sign);

        // if (number == INT_MAX && sign == -1) {
        //     return INT_MIN;
        // }

        return number;

    }
};